source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -343,7 +350,11 @@ void Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark() {
 }
 
 void Filter::UpstreamCallbacks::onBelowWriteBufferLowWatermark() {
-  ASSERT(on_high_watermark_called_);
+  // TCP Tunneling may call on high watermark multiple times.
+  if (!parent_->config_->tunnelingConfigHelper()) {
+    ASSERT(on_high_watermark_called_);
```

Comment:
Could consider putting the logic of the if into the ASSERT.

e.g. `ASSERT(parent_->config-_->tunnelingConfigHelper() || (!parent_->config_->tunnelingConfigHelper() && on_high_watermark_called_))` which would ensure the entire check is compiled out in optimized mode (as currently done)

Reply:
done